controllers/rollout/trafficrouting.go

```diff
@@ -0,0 +1,283 @@
+			r.rollout.Namespace, r.rollout.Name, r.stableService.Name, r.podRevisionLabelKey(), r.newStatus.StableRevision)
+		return false, nil
+	}
+
+	// After restore stable service configuration, give the ingress provider 3 seconds to take effect
+	if verifyTime := canaryStatus.LastUpdateTime.Add(time.Second * 3); verifyTime.After(time.Now()) {
+		klog.Infof("update rollout(%s/%s) stable service(%s) done, and wait 3 seconds", r.rollout.Namespace, r.rollout.Name, r.stableService.Name)
+		return false, nil
+	}
+
+	// route traffic configuration
+	trController, err := r.newTrafficRoutingController(r)
+	if err != nil {
+		klog.Errorf("rollout(%s/%s) newTrafficRoutingController failed: %s", r.rollout.Namespace, r.rollout.Name, err.Error())
+		return false, err
+	}
+	var desiredWeight int32
+	if len(r.rollout.Spec.Strategy.Canary.Steps) > 0 {
+		desiredWeight = r.rollout.Spec.Strategy.Canary.Steps[r.newStatus.CanaryStatus.CurrentStepIndex-1].Weight
+	}
+	steps := len(r.rollout.Spec.Strategy.Canary.Steps)
+	cond := util.GetRolloutCondition(*r.newStatus, appsv1alpha1.RolloutConditionProgressing)
+	cond.Message = fmt.Sprintf("Rollout is in step(%d/%d), and route traffic weight(%d)", canaryStatus.CurrentStepIndex, steps, desiredWeight)
+	verify, err := trController.VerifyTrafficRouting(desiredWeight)
+	if err != nil {
+		return false, err
+	} else if !verify {
+		return false, trController.SetRoutes(desiredWeight)
+	}
+	klog.Infof("rollout(%s/%s) do step(%d) trafficRouting(%d%) success", r.rollout.Namespace, r.rollout.Name, r.newStatus.CanaryStatus.CurrentStepIndex, desiredWeight)
+	return true, nil
+}
+
+func (r *rolloutContext) restoreStableService() (bool, error) {
+	//fetch stable service
+	sName := r.rollout.Spec.Strategy.Canary.TrafficRouting.Service
+	r.stableService = &corev1.Service{}
+	err := r.Get(context.TODO(), client.ObjectKey{Namespace: r.rollout.Namespace, Name: sName}, r.stableService)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+		klog.Errorf("rollout(%s/%s) get stable service(%s) failed: %s", r.rollout.Namespace, r.rollout.Name, sName, err.Error())
+		return false, err
+	}
+
+	if r.newStatus.CanaryStatus == nil {
+		r.newStatus.CanaryStatus = &appsv1alpha1.CanaryStatus{}
+	}
+	//restore stable service configuration，remove hash revision selector
+	if r.stableService.Spec.Selector != nil && r.stableService.Spec.Selector[r.podRevisionLabelKey()] != "" {
+		delete(r.stableService.Spec.Selector, r.podRevisionLabelKey())
+		if err := r.retryUpdateService(r.stableService); err != nil {
+			return false, err
+		}
+		klog.Infof("remove rollout(%s/%s) stable service(%s) pod revision selector success, and retry later", r.rollout.Namespace, r.rollout.Name, r.stableService.Name)
+		r.newStatus.CanaryStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+		return false, nil
+	}
+	// After restore stable service configuration, give the ingress provider 3 seconds to take effect
```

Comment:
how to confirm ingress is ready in 3 seconds

Reply:
This place is currently the default 3seconds, and GracePeriodSeconds has now been added to user-defined.